File: src/gcs/examples.py
"""Reference sketches used by the tests, the benchmarks and the app's case library.

The sketches themselves are built in the core; this is the lookup and the case list.
"""

from __future__ import annotations

from typing import Any, Callable

from gcs import _ffi
from gcs._ffi import lib
from gcs.model import Sketch


def build(name: str) -> Sketch:
    p, n = _ffi.send(name)
    h = lib.gcs_example(p, n)
    if not h:
        raise KeyError(_ffi.last_error() or name)
    return Sketch(h)


def rect_fillets(w: float = 100.0, h: float = 60.0, r: float = 10.0) -> Sketch:
    return build(f"rect_fillets:{w}:{h}:{r}")


def slotted_link(length: float = 80.0, r: float = 15.0, hole_r: float = 6.0) -> Sketch:
    return build("slotted_link")


def truss(bays: int = 8) -> Sketch:
    return build(f"truss:{bays}")


def polygon_chain(n: int = 12) -> Sketch:
    return build(f"polygon_chain:{n}")


def rect_fillets_conflict() -> Sketch:
    return build("rect_fillets_conflict")


def rect_fillets_under() -> Sketch:
    return build("rect_fillets_under")


def truss_redundant() -> Sketch:
    return build("truss_redundant")


def truss_conflict() -> Sketch:
    return build("truss_conflict")


def truss_floating(bays: int = 8) -> Sketch:
    return build(f"truss_floating:{bays}")


def zigzag(n: int = 32, copies: int = 1) -> Sketch:
    """`copies` separate staircases of `n` points, every segment alternately vertical and
    horizontal and nothing else: the sketch that finds any cost going with the document rather
    than with the figure being dragged."""
    return build(f"zigzag:{n}:{copies}")


def impossible_triangle() -> Sketch:
    return build("impossible_triangle")


def altitudes() -> Sketch:
    return build("altitudes")


def belt_tangency() -> Sketch:
    """A belt over two pulleys, each end held on its circle and the line tangent to it: a double
    root — rank-deficient at every solution, yet nothing can move."""
    return build("belt_tangency")


def parallels() -> Sketch:
    return build("parallels")


def pythagoras(a: float = 30.0, b: float = 40.0) -> Sketch:
    """The graphical proof of the Pythagorean theorem: four a×b right triangles in a square of
    side a + b leave a square whose side is claimed to be `c = hypot(a, b)` — judged a
    theorem."""
    return build(f"pythagoras:{a}:{b}")


def k33() -> Sketch:
    """A K3,3 bar framework: rigid, triangle-free, and so one core to the decomposition."""
    return build("k33")


def laman(n: int = 10, seed: int = 0) -> Sketch:
    return build(f"laman:{n}:{seed}")


def spline_follower() -> Sketch:
    """A cubic B-spline with a face held tangent to it and a point riding on it.

    The control points are written out in the document, so how many there are is its own
    and not an argument.
    """
    return build("spline_follower")


def peaucellier() -> Sketch:
    """The Peaucellier–Lipkin cell: circling rods whose pen draws an exact straight line, the
    straightness one `claim` the diagnosis judges a theorem."""
    return build("peaucellier")


def perturb(sk: Sketch, sigma: float, seed: int = 0) -> None:
    sk.perturb(sigma, seed)


def henneberg_edges(n: int, seed: int = 0) -> list[tuple[int, int]]:
    """A random Laman graph on n vertices by Henneberg I/II moves — minimally rigid by
    construction, and the generator the property tests use."""
    e = _ffi.take_json(lib.gcs_henneberg_edges_json(int(n), seed & 0xFFFFFFFF)) or []
    return [(a, b) for a, b in e]


EXAMPLES: dict[str, Callable[[], Sketch]] = {
    "rect_fillets": rect_fillets,
    "slotted_link": slotted_link,
    "truss": truss,
    "polygon_chain": polygon_chain,
    "spline_follower": spline_follower,
}


def cases() -> list[dict[str, Any]]:
    """The case library shown in the app: label, key and a one-line description."""
    return _ffi.take_json(lib.gcs_cases_json()) or []


__all__ = ["EXAMPLES", "altitudes", "henneberg_edges", "build", "cases", "impossible_triangle", "k33", "laman", "pythagoras",
           "parallels", "peaucellier", "perturb", "polygon_chain", "rect_fillets", "rect_fillets_conflict",
           "rect_fillets_under", "slotted_link", "truss", "truss_conflict", "truss_floating",
           "belt_tangency",
           "spline_follower", "truss_redundant", "zigzag"]

File: web/src/core/examples.ts
/* Reference sketches used by the tests, the benchmarks and the app's case library.
 *
 * The sketches themselves are built in the core; this is the lookup and the case list. */
import { Sketch } from './model.js';
import { fromSketch } from './program.js';
import { core, lastError, takeJson, takeStr, withStr } from './wasm.js';

/** Build a named example.  Keys are a plain name or `name:arg[:arg]` — `truss:50`, `laman:12:1`. */
export function build(key: string): Sketch {
  const h = withStr(key, (p, n) => core().gcs_example(p, n));
  if (!h) throw new Error(lastError() || `unknown example: ${key}`);
  return new Sketch(h);
}

/** A named example as a *program* — which is what the document is.
 *
 *  A case written as a document has a source somebody wrote, comments and components and all, and
 *  that source is the case.  One built by a function has none, so it is lifted; either way the
 *  caller gets a program and never has to know which kind it was. */
export function source(key: string): string {
  const h = withStr(key, (p, n) => core().gcs_example_source(p, n));
  if (h) return takeStr(h);
  const sk = build(key);
  try {
    return fromSketch(sk);
  } finally {
    sk.dispose();
  }
}

export const rectFillets = (w = 100, h = 60, r = 10): Sketch => build(`rect_fillets:${w}:${h}:${r}`);
export const slottedLink = (): Sketch => build('slotted_link');
export const truss = (bays = 8): Sketch => build(`truss:${bays}`);
export const polygonChain = (n = 12): Sketch => build(`polygon_chain:${n}`);
export const rectFilletsConflict = (): Sketch => build('rect_fillets_conflict');
export const rectFilletsUnder = (): Sketch => build('rect_fillets_under');
export const trussRedundant = (): Sketch => build('truss_redundant');
export const trussConflict = (): Sketch => build('truss_conflict');
export const trussFloating = (bays = 8): Sketch => build(`truss_floating:${bays}`);
export const impossibleTriangle = (): Sketch => build('impossible_triangle');
export const altitudes = (): Sketch => build('altitudes');
/** A belt over two pulleys, each end held on its circle and the line tangent to it: a double
 *  root — rank-deficient at every solution, yet nothing can move. */
export const beltTangency = (): Sketch => build('belt_tangency');
export const parallels = (): Sketch => build('parallels');
/** The graphical proof of the Pythagorean theorem: four a×b right triangles in a square of side
 *  a + b leave a square whose side is claimed to be `c = hypot(a, b)` — judged a theorem. */
export const pythagoras = (a = 30, b = 40): Sketch => build(`pythagoras:${a}:${b}`);
export const k33 = (): Sketch => build('k33');
export const laman = (n = 10, seed = 0): Sketch => build(`laman:${n}:${seed}`);
/** A cubic B-spline with a face held tangent to it and a point riding on it. */
/** The control points are written out in the document, so their number is its own. */
export const splineFollower = (): Sketch => build('spline_follower');
/** `copies` separate staircases of `n` free-length H/V segments: a drag must cost one figure. */
export const zigzag = (n = 32, copies = 1): Sketch => build(`zigzag:${n}:${copies}`);
/** The Peaucellier–Lipkin cell: circling rods whose pen draws an exact straight line, the
 *  straightness one `claim` the diagnosis judges a theorem. */
export const peaucellier = (): Sketch => build('peaucellier');

/** A random Laman graph's edges by Henneberg construction — the property-test generator. */
export function hennebergEdges(n: number, seed = 0): [number, number][] {
  return takeJson<[number, number][]>(core().gcs_henneberg_edges_json(n, seed >>> 0)) ?? [];
}

export const EXAMPLES: Record<string, () => Sketch> = {
  rect_fillets: () => rectFillets(),
  slotted_link: () => slottedLink(),
  truss: () => truss(),
  polygon_chain: () => polygonChain(),
  spline_follower: () => splineFollower(),
};

export interface Case {
  label: string;
  key: string;
  description: string;
}

/** The case library shown in the app: label, key and a one-line description. */
export function cases(): Case[] {
  return takeJson<Case[]>(core().gcs_cases_json()) ?? [];
}
